src/extract_gpuinfo_tpu.c

```diff
@@ -0,0 +1,290 @@
+double* _duty_cycle_pct;
+
+bool is_cache_valid(void) {
+  nvtop_time current_time;
+  nvtop_get_current_time(&current_time);
+  uint64_t t_diff_ns = nvtop_difftime_u64(last_cache_refresh, current_time);
+  return t_diff_ns < 900 * 1000 * 1000; // 900ms
+}
+
+bool refresh_tpu_cache(void) {
+  if (is_cache_valid()) return true;
+  nvtop_get_current_time(&last_cache_refresh);
+  if (tpu_chip_count <= 0) return false;
+  if (_tpu_pids(_pids, tpu_chip_count) != 0) {
+    reset_tpu_cache(false);
+    return false;
+  }
+  for (int i = 0; i < tpu_chip_count; i++) latest_chips_usage_data[i].pid = _pids[i];
+
+  if (_tpu_metrics(tpu_runtime_monitoring_port, _device_ids, _memory_usage, _total_memory,
+                   _duty_cycle_pct, tpu_chip_count) != 0) return false;
+  for (int i = 0; i < tpu_chip_count; i++) {
+    latest_chips_usage_data[i].device_id = _device_ids[i];
+    latest_chips_usage_data[i].memory_usage = _memory_usage[i];
+    latest_chips_usage_data[i].total_memory = _total_memory[i];
+    latest_chips_usage_data[i].duty_cycle_pct = _duty_cycle_pct[i];
+  }
+  return true;
+}
+
+void reset_tpu_cache(bool fully) {
+  for (int i = 0; i < tpu_chip_count; i++) {
+    latest_chips_usage_data[i].memory_usage = 0;
+    latest_chips_usage_data[i].duty_cycle_pct = 0;
+    latest_chips_usage_data[i].pid = -1;
+    if (fully) {
+      snprintf(latest_chips_usage_data[i].name, sizeof(latest_chips_usage_data[i].name), "%s", "N/A");
+      latest_chips_usage_data[i].device_id = 0;
+      latest_chips_usage_data[i].total_memory = 0;
+    }
+  }
+}
+/* TPU info cache ------------------------------------------------------------------------------- */
+
+bool gpuinfo_tpu_init(void) {
+  char* error_msg;
+  char symbol_name[64];
+  char library_path[256];
+  nvtop_get_current_time(&last_cache_refresh);
+  // invalidate cache by putting it in the past
+  last_cache_refresh = nvtop_substract_time(last_cache_refresh, (nvtop_time){10, 0}); 
+
+  // Load dynamic library symbols
+  snprintf(library_path, sizeof(library_path), "%s", libname);
+  void *handle = dlopen(library_path, RTLD_LAZY);
+  if (!handle) {
+      error_msg = dlerror();
+      if (error_msg != NULL) fprintf(stderr, "TPU support error: %s\n", error_msg);
+      return false;
+  }
+  snprintf(symbol_name, sizeof(symbol_name), "%s", "tpu_chip_count");
+  _tpu_chip_count = dlsym(handle, "tpu_chip_count");
+  error_msg = dlerror();
+  if (error_msg != NULL) goto dlsymb_cleanup;
+  snprintf(symbol_name, sizeof(symbol_name), "%s", "tpu_pids");
+  _tpu_pids = dlsym(handle, "tpu_pids");
+  error_msg = dlerror();
+  if (error_msg != NULL) goto dlsymb_cleanup;
+  snprintf(symbol_name, sizeof(symbol_name), "%s", "tpu_metrics");
+  _tpu_metrics = dlsym(handle, "tpu_metrics");
+  error_msg = dlerror();
+  if (error_msg != NULL) goto dlsymb_cleanup;
+
+  // Discover TPU devices
+  tpu_chip_count = _tpu_chip_count();
+  if (tpu_chip_count == 0) {
+    fprintf(stderr, "Found 0 TPU devices on the system.\n");
```

Comment:
The printfs usually do not coexist well with the ncurses interface, better to remove them or put them behind a debug flag

Reply:
Added an NDEBUG guard